### PR TITLE
Improve JavaScript doc comments for generated code

### LINKED ISSUE #2409: Invalid link generated in JavaScript doc comment
We generate

```
/**
 * The number of threads which are currently performing other activities. These are all other that are not
 * counted with {@link #inUseForUser} or {@link #inUseForIO}, such as DNS lookups, garbage collection).
 */
 inUseForOther: number;
```

TypeDoc fails to resolve those links:

```
npx typedoc dist/bundle.d.ts --disableGit --disableSources
[warning] Failed to resolve link to "#inUseForUser" in comment for IceMX.ThreadMetrics.inUseForOther
[warning] Failed to resolve link to "#inUseForIO" in comment for IceMX.ThreadMetrics.inUseForOther
```

We should generate `{@link ClassName#methodName}` instead.

### PR DESCRIPTION
**Pull request by pepone**
Fix #2409

### CODE REVIEW

File: cpp/src/Slice/Parser.cpp [hunk 1]
```diff
@@ -732,6 +732,12 @@ namespace
 
 CommentPtr
 Slice::Contained::parseComment(bool stripMarkup) const
+{
+    return parseComment(_comment, stripMarkup);
+}
+
+CommentPtr
+Slice::Contained::parseComment(const string& text, bool stripMarkup) const
```
Comment (pepone) [inline cpp/src/Slice/Parser.cpp:740]: I added this overload to be able to preprocess the coment before parsing. This allow for language specific preprocessing.

File: cpp/src/slice2js/Gen.cpp [hunk 2]
```diff
@@ -252,7 +284,6 @@ namespace
         const OperationPtr& op,
         const CommentPtr& doc,
         OpDocParamType type,
-        const StringList& preParams = StringList(),
```
Comment (pepone) [inline cpp/src/slice2js/Gen.cpp:255]: Remove this param, we were always passing an empty list.

File: cpp/src/slice2js/Gen.cpp [hunk 3]
```diff
@@ -303,9 +329,10 @@ namespace
             ExceptionPtr ex = op->container()->lookupException(name, false);
             if (ex)
             {
-                name = ex->scoped().substr(2);
+                name = ex->scoped();
             }
-            out << nl << " * @throws " << name << " ";
+            name = JsGenerator::fixId(name);
+            out << nl << " * @throws {@link " << name << "} ";
             writeDocLines(out, p->second, false);
```
Comment (pepone) [inline cpp/src/slice2js/Gen.cpp:336]: TypeDoc uses this syntax in throws see: https://typedoc.org/tags/throws/

File: cpp/src/slice2js/Gen.cpp [hunk 4]
```diff
@@ -315,8 +342,6 @@ namespace
         const OperationPtr& op,
         const CommentPtr& doc,
         OpDocParamType type,
-        bool showExceptions,
```
Comment (pepone) [inline cpp/src/slice2js/Gen.cpp:318]: Removed, we should always write the @throws clauses. We were passing "false" all the time and exceptions were ignored.

File: cpp/src/slice2js/Gen.cpp [hunk 5]
```diff
@@ -22,6 +22,38 @@ using namespace IceInternal;
 
 namespace
 {
+    CommentPtr parseComment(const ContainedPtr& p)
+    {
+        // JavaScript TypeDoc doc processor doesn't accept # at the beginning of a link
+        // so we need to remove it.
```
Comment (externl) [inline cpp/src/slice2js/Gen.cpp:28]: Do these generate valid links still?

**Review by externl — APPROVED** [review]

File: cpp/src/slice2js/Gen.cpp (hunk 5 above)
Comment (pepone) [reply cpp/src/slice2js/Gen.cpp:28]: Yes, with Slice and Java doc you can use `{@link #foo} to refer to a member of the class. With TypeDoc is just `{@link foo}`

This was causing warnings mentioned above when parasing https://github.com/zeroc-ice/ice/blob/5d4f24b67aed1cb70412c54dd03035f4179582c4/slice/Ice/Metrics.ice#L134-L136

**Review by bernardnormier — APPROVED** [review]

**Pull request merged**